src/System.Net.Security/src/System/PinnableBufferCache.cs

```diff
@@ -0,0 +1,552 @@
+                {
+                    PinnableBufferCacheEventSource.Log.AllocateBufferFromNotGen2(_cacheName, _notGen2.Count);
+                }
+
+                // If we have a Gen1 collection, then everything on m_NotGen2 should have aged.  Move them to the m_Free list.  
+                if (!AgePendingBuffers())
+                {
+                    // Before we could age at set of buffers, we have handed out half of them.
+                    // This implies we should be proactive about allocating more (since we will trim them if we over-allocate).  
+                    if (_notGen2.Count == _restockSize / 2)
+                    {
+                        PinnableBufferCacheEventSource.Log.DebugMessage("Proactively adding more buffers to aging pool");
+                        CreateNewBuffers();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// See if we can promote the buffers to the free list.  Returns true if successful. 
+        /// </summary>
+        [System.Security.SecuritySafeCritical]
+        private bool AgePendingBuffers()
+        {
+            if (_gen1CountAtLastRestock < GC.CollectionCount(GC.MaxGeneration - 1))
+            {
+                // Allocate a temp list of buffers that are not actually in gen2, and swap it in once
+                // we're done scanning all buffers.
+                int promotedCount = 0;
+                List<object> notInGen2 = new List<object>();
+                PinnableBufferCacheEventSource.Log.AllocateBufferAged(_cacheName, _notGen2.Count);
+                for (int i = 0; i < _notGen2.Count; i++)
+                {
+                    // We actually check every object to ensure that we aren't putting non-aged buffers into the free list.
+                    object currentBuffer = _notGen2[i];
+                    if (GC.GetGeneration(currentBuffer) >= GC.MaxGeneration)
+                    {
+                        _freeList.Push(currentBuffer);
+                        promotedCount++;
+                    }
+                    else
+                    {
+                        notInGen2.Add(currentBuffer);
+                    }
+                }
+                PinnableBufferCacheEventSource.Log.AgePendingBuffersResults(_cacheName, promotedCount, notInGen2.Count);
+                _notGen2 = notInGen2;
+
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Generates some buffers to age into Gen2.
+        /// </summary>
+        private void CreateNewBuffers()
+        {
+            // We choose a very modest number of buffers initially because for the client case.  This is often enough.
+            if (_restockSize == 0)
+                _restockSize = 4;
+            else if (_restockSize < DefaultNumberOfBuffers)
+                _restockSize = DefaultNumberOfBuffers;
+            else if (_restockSize < 256)
```

Comment:
Won't fix. The magic numbers are localized within the function that's using them. Tuning cannot be done without changing this function (note the fractional restock growth rate for large restock sizes). 

I believe the code as written is cleaner.